src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraph.java

```diff
@@ -0,0 +1,464 @@
+            return out.updateCachesFromAccumulatedValues(accumulatedValues);
+        } else {
+            return out;
+        }
+    }
+
+    /**
+     * Make a key -> node map containing new instances of the nodes which become out of date as a matter of
+     * updating node {@code key}
+     * @param key key of the updated note
+     * @param updatedNodesMap a key -> node map
+     */
+    private Map<String, CacheNode> addDuplicateOfOutdatedDescendents(@Nonnull final String key,
+                                                                     @Nonnull final Map<String, CacheNode> updatedNodesMap) {
+        for (final String child : cgs.getAllDescendentsMap().get(key)) {
+            CacheNode oldChild = nodesMap.get(child);
+            /* all of the desendents are computable nodes and can be safely upcasted */
+            updatedNodesMap.put(child, ((ComputableCacheNode)oldChild).duplicateWithUpdatedCacheStatus(false));
+        }
+        return updatedNodesMap;
+    }
+
+    /**
+     * Returns a reference to the value of a given node
+     *
+     * Note: this function is purposefully meant to be light:
+     *
+     * (1) it does not update out-of-date cacheable computable nodes, and
+     * (2) it does not evaluate non-cacheable computable nodes.
+     *
+     * @param nodeKey the key of the node
+     * @return value of the node
+     * @throws IllegalStateException if a cached node is out of date, or if the value of a primitive node is
+     *                               not initialized
+     * @throws IllegalArgumentException if the node does not exist
+     */
+    public Duplicable getValueDirect(@Nonnull final String nodeKey)
+            throws IllegalStateException, IllegalArgumentException {
+        assertNodeExists(Utils.nonNull(nodeKey, "The key of the node can not be null."));
+        return nodesMap.get(nodeKey).getValue(EMPTY_MAP);
+    }
+
+    /**
+     *
+     * @param nodeKey key of the node
+     * @return value of the node
+     * @throws IllegalArgumentException if the node does not exist
+     */
+    public Duplicable getValueWithRequiredEvaluations(@Nonnull final String nodeKey)
+            throws IllegalStateException, IllegalArgumentException {
+        assertNodeExists(Utils.nonNull(nodeKey, "The key of the node can not be null."));
+        return accumulateEvaluationsForDepthSortedNodeList(cgs.getEvalStrategyByEvaluatedNode(nodeKey), true).get(nodeKey);
+    }
+
+    /**
+     * This method calculates and accumulates the values of a list of depth-sorted nodes (from low depth to high depth).
+     * Compute-on-demand nodes are always evaluated. Caching nodes are only evaluated if {@code evalOnOutOfDateCachingNodes}
+     * is true.
+     *
+     * @param depthSortedNodes depth-sorted list of nodes
+     * @param evalOutdatedCachingNodes evaluate
+     * @return
+     */
+    private Map<String, Duplicable> accumulateEvaluationsForDepthSortedNodeList(@Nonnull final List<String> depthSortedNodes,
```

Comment:
While I was at it, I (1) created a new functional interface `ComputableNodeFunction` and got rid of the ugly and highly repeated `Function<Map<String, ? extends Duplicable>, ? extends Duplicable>`, and (2) created a whole bunch of new ICG-specific Exceptions.